### PR TITLE
DM-45522: `PydanticSchemaManager` for confluent schema registry interactions

### PR DESCRIPTION
**Pull request by fajpunk**
Basically Jonathon's [kafkit](https://kafkit.lsst.io/guide/index.html) pydantic schema manager with the hard parts replaced by [python-schema-registry-client](https://marcosschroh.github.io/python-schema-registry-client).

### CODE REVIEW

**Comment by jonathansick** [comment]
Given the broader scope of safir, I'm not sure the `schema_manager` module name sits well. Do you think this should be a feature of the Kafka module? Or if we want to believe that the Confluent Schema Registry is useful beyond Kafka, do you think the `schema_manager` name should be more specific? In other words, I'd vote for this in order of preference:

- `from safir.kafka import PydanticSchemaRegistryManager` (treat as a Kafka feature)
- `from safir.schemaregistry import PydanticSchemaRegistryManager` (treat as a client to the Schema Registry, like the safir.arq and safir.github subpackages)

Curious to hear what you and @rra think.

**Comment by fajpunk** [comment]
Yeah, I started all this work with this in the Kafka module, and went through a lot of internal hemming and hawing before moving it out.  The schema registry theoretically could be used using Kafka. In practice, I assume we don't currently ever use it like that, and I can't think of anything concrete coming up where we would.

Given that, having it be part of `safir.kafka` is my vote. Maybe we want to split that up into submodules in that case, so that namespace doesn't get too crowded, maybe `kafka.config` and `kafka.schema_registry` (so I guess a combination of both of your suggestions :) )?

File: safir/src/safir/schema_manager/_pydantic_schema_manager.py
Comment (rra) [inline safir/src/safir/schema_manager/_pydantic_schema_manager.py:206]: The typing here is pretty ugly. I feel like there should be some better way to do this, but I'd have to do some research to see if there's some way to properly type this. The nested subclass thing is kind of a nightmare for typing, which is why Pydantic moved away from it in v2. Some judicious use of `getattr` might help here?

I hate to have run-time type checks when the whole point of mypy is to do that during development and not have to write explicit code to do them, but there may just not be a way around that if mypy can't understand that certain classes have embedded classes that implement a particular protocol (other than blindly stringifying whatever we get, which honestly wouldn't be that unreasonable).

**Review by rra — COMMENTED** [review]
Personally, I'd put this all in with `safir.kafka` since in the Safir context it's specific to the Safir Kafka support, and I would also be inclined to just export all the symbols from the root `safir.kafka` module and not have submodules. I don't have a strong opinion about this, and younger me would have disagreed, but the more Python libraries I use, the more I appreciate not having to think about someone else's module structure and just being able to import all the stuff from the top level.

If we do keep a submodule, I might make a mild argument for `schema` over `schema_manager` because the latter still feels awfully verbose, but at that point I'm probably just being irrational. :)

File: safir/src/safir/schema_manager/_pydantic_schema_manager.py
Comment (rra) [reply safir/src/safir/schema_manager/_pydantic_schema_manager.py:57]: Oh, this is an internal object, never mind. I usually write doc strings anyway, but if it's not exported, it doesn't really matter if you feel like it's obvious in context.

**Comment by rra** [comment]
Oh, the other thing that I was kind of wanting was some more Safir-specific integration documentation for exactly how to set up the schema manager. There's enough information there that I can work out something reasonable, but maybe the documentation should suggest a specific approach? For example, all the schemas should be registered, probably during startup, so presumably that should happen in the `lifespan` hook for the FastAPI app? There should be a shared schema manager used for serializing things, so maybe it should be a dependency singleton? (Should we provide one as `safir.dependencies.kafka.kafka_schema_manager` or something like that?)

File: docs/user-guide/pydantic-schema-manager.rst
Comment (fajpunk) [reply docs/user-guide/pydantic-schema-manager.rst:109]: I'm actually totally with you on this, I was trying to match style elsewhere in the repo, like https://github.com/lsst-sqre/safir/blob/main/safir/src/safir/github/models.py

File: safir/src/safir/schema_manager/_pydantic_schema_manager.py
Comment (fajpunk) [reply safir/src/safir/schema_manager/_pydantic_schema_manager.py:149]: This is here just to satisfy the type checker. The [test_compatibility](https://github.com/marcosschroh/python-schema-registry-client/blob/09ab31bbdf7bda776e5d344ccf277af9b1350881/schema_registry/client/client.py#L670) method returns a `bool` if `verbose == False`, and a `dict` if `verbose == True`. They could've used `@overload`s or something to make it typecheck nicely, but they didn't :)

Is there a nicer way to satisfy the typechecker here?  Should I just have the error message say "This is only here to satisfy the type checker?"

File: safir/src/safir/schema_manager/_pydantic_schema_manager.py
Comment (fajpunk) [reply safir/src/safir/schema_manager/_pydantic_schema_manager.py:150]: This is another case where the `pydantic-schema-manager` lib is not nicely typed, and my instincts would be to check for the key and it's not there, raise some exception like "There's something wrong with this 3rd party lib, this should never happen"

Similar to https://github.com/lsst-sqre/safir/pull/303#discussion_r1776205987

File: safir/src/safir/schema_manager/_pydantic_schema_manager.py
Comment (fajpunk) [reply safir/src/safir/schema_manager/_pydantic_schema_manager.py:206]: I did have a way to typecheck this... I don't remember exactly what I did, but it was pretty nasty, I think it did do a bunch of `getattr`. I eventually decided to leave this as is, because getting it to a place where I could get rid of the `# type: ignore` didn't make anything any safer and made the code much more confusing.

I'd be glad to dig it up again, but I'd like one more assurance that you really want to see it :)

Comment (fajpunk) [reply safir/src/safir/schema_manager/_pydantic_schema_manager.py:206]: ...and I missed the whole point of your comment and you're talking about the `isinstance` checks of the `Meta` attributes.  Yeah, I'm not sure if there's a better way to do that, but I like the idea of just stringifying whatever they contain.  I'll do that if I can't find some fancy way of making that check static.

**Comment by fajpunk** [comment]
Right now, the schema manager is only going to be used inside the metrics EventManager (see [this undocumented-right-now PR](https://github.com/lsst-sqre/safir/pull/304/files) for how). In an app that uses the schema manager directly, I imagine it would work exactly like you postulate: it would be a process-scoped singleton dependency where all the schemas get registered in a `lifespan` hook.  That's how I'm planning on the `EventManager` working.

File: docs/user-guide/pydantic-schema-manager.rst
Comment (fajpunk) [reply docs/user-guide/pydantic-schema-manager.rst:188]: > Here also having lots of Python code blocks right next to each other feels a bit confusing to me, and I'd instead lift the comments into regular text outside of the code blocks.

I think this ends up being more confusing in that it's difficult to know which text goes with which code block.  Is there a way you would recommend to make that more clear?
image

File: docs/user-guide/pydantic-schema-manager.rst
Comment (fajpunk) [reply docs/user-guide/pydantic-schema-manager.rst:202]: I don't think it makes sense to tie it to `config.profile`, assuming our standard practice is to set that to "development" in the Phalanx dev environments, and "production" everywhere else.  In that case, there would still be the same subject for any model deployed to one of those environments in that environment's Schema Registry, it would just be different than that model's subject in the non-dev environments' schema managers. You could still clobber the compatibility in a `config.profile = development` environment by not setting a _different_ subject when you point your phalanx app to a different branch that has model changes that don't make it back into the main branch.

If we wanted to do something automatic, I think we'd want something like, in dev envs, the suffix is the branch name.  In that case though, whatever is on the other end, presumably a different phalanx app, would have to deal with all of the different subjects.

In reality, I think we're likely to screw up compatibility in the registry in a dev env, and we can "just" delete the schema or specific versions of it from the registry when that happens. It stinks, and I don't immediately know how to do any better than "here's a think you can use to try to avoid that if you can remember it" :(

I think that pain is partly why @jonathansick moved away from using the Schema Registry (and to shared Pydantic models) when we control both ends, but I don't want to put words in his mouth, and he may have some more useful thoughts than me on this.

Comment (fajpunk) [reply docs/user-guide/pydantic-schema-manager.rst:202]: I'm totally wrong, `suffix` has no effect on the Avro record name, and I don't think anyone consuming the encoded messages cares about what subject they're in in the Schema Registry.

File: safir/src/safir/schema_manager/_config.py
Comment (fajpunk) [reply safir/src/safir/schema_manager/_config.py]: At this point, there's no case where anyone would want to use the registry without going though the manager. I combined the settings.

To that point, I also made a convenience method on the settings class to make a `PydanticSchemaManager`, because there's no way to construct one any other way from the settings.

File: safir/src/safir/schema_manager/_pydantic_schema_manager.py
Comment (fajpunk) [reply safir/src/safir/schema_manager/_pydantic_schema_manager.py]: Yeah, I ended up back at a bunch of nested `getattr`s, so I just `str`'d the attribute and got rid of the isinstance checks: https://github.com/lsst-sqre/safir/pull/303/commits/0faa5fd4a6e42e0e2e631261bbc64dbe3e5037dc

If you like nested getattrs better, I'll change it.

File: docs/user-guide/pydantic-schema-manager.rst
Comment (rra) [reply docs/user-guide/pydantic-schema-manager.rst]: I have a whole complicated set of mental heuristics for this that are probably too complicated to write down and thus are probably unrealistically complicated, heh. Yeah, I'd probably use the longer names for the GitHub models because classes like that feel different to me. I think the distinction is how much context there is at the call site. Class names feel like they're often used in situations with less context, and also they're often multi-line blocks anyway so the penalty for longer names is less. Enums are often used as arguments to functions or methods where the context is really obvious, so I tend to make them shorter. This is absolutely me overthinking all of this.

File: docs/user-guide/pydantic-schema-manager.rst
Comment (rra) [reply docs/user-guide/pydantic-schema-manager.rst]: Ah, okay, yeah, `config.profile` doesn't seem to make sense.

I think I'm just confused about how I should decide when to add a prefix and when to remove it again and the process seems somewhat error-prone. It sounds like that's somewhat inherent in the problem space, though, and we may not be able to avoid it.

Could it make sense to use a compatibility setting for the dev environment that allows people to push entirely incompatible schemas, and then use a tighter one in the integration environment? And then really stress in the documentation that you must finalize your schema before you deploy on integration?

File: docs/user-guide/pydantic-schema-manager.rst
Comment (rra) [reply docs/user-guide/pydantic-schema-manager.rst]: How about keeping the consecutive code blocks but moving the comment into a `:caption:` attribute on the code-block directive? I think that will add the visual separation that I'm wanting but avoid the ambiguity problem.

**Comment by rra** [comment]
Ah, okay, sounds like it's best to hold off on deeper FastAPI-aimed documentation on how to use this since mostly people will use the event manager instead. I'm still not sure if I'm going to use this directly in Gafaelfawr or if the event manager will do everything that I need to do Kafka-wise.

File: safir/src/safir/schema_manager/_config.py
Comment (rra) [reply safir/src/safir/schema_manager/_config.py]: Oh, great, thank you!

File: safir/src/safir/schema_manager/_pydantic_schema_manager.py
Comment (rra) [reply safir/src/safir/schema_manager/_pydantic_schema_manager.py]: Because it's not properly typed upstream, I am still a bit worried that we might miss an upstream semantic change and the exception might actually fire, so I'm not sure it's only to satisfy the type checker. (It's checking that the contract hasn't changed because the type checker can't do that for us because upstream doesn't have proper typing, or at least that's how I think about it.)

Maybe something like:
```python
if not isinstance(result, dict) or "is_compatible" not in result:
    raise RuntimeError(f"Unexpected result from test_compatibility: {result!s}")
```

File: docs/user-guide/pydantic-schema-manager.rst
Comment (fajpunk) [reply docs/user-guide/pydantic-schema-manager.rst]: I shortened some other names in the kafka connection stuff too:
https://github.com/lsst-sqre/safir/pull/303/commits/617b8daa65454ed9b16622cf05917f2719db2587

File: safir/src/safir/schema_manager/_config.py
Comment (fajpunk) [reply safir/src/safir/schema_manager/_config.py]: I did this and changed some of the methods in the kafka config stuff to be consistent:
https://github.com/lsst-sqre/safir/pull/303/commits/57e1c59e9e51fafbc0dacb86dd33861cff8051f0

File: safir/src/safir/schema_manager/_pydantic_schema_manager.py
Comment (rra) [reply safir/src/safir/schema_manager/_pydantic_schema_manager.py]: Yeah, that seems fine. I think that's as good as we can do. https://github.com/python/mypy/issues/15835 is the feature that I really want here, but it appears to be a hard problem and no one has found a good solution yet.

**Comment by fajpunk** [comment]
Holding off on on deeper FastAPI-aimed documentation for the schema manager sounds good to me for now.

And yep, for apps like Gafaelfawr that don't already have a kafka integration, the EventManager will do everything you need kafka-wise.

File: docs/user-guide/pydantic-schema-manager.rst
Comment (fajpunk) [reply docs/user-guide/pydantic-schema-manager.rst]: > And then really stress in the documentation that you must finalize your schema before you deploy on integration?

This for sure!

> Could it make sense to use a compatibility setting for the dev environment that allows people to push entirely incompatible schemas, and then use a tighter one in the integration environment?

I was initially against this because it would be nice to find out if your schemas are incompatible before you get to int (and thus likely after your changes have been merged), but in a previous comment I said "we're likely to screw it up in dev and have to delete subjects and versions", and if we did that, then it's not any better than if we had the NONE compatibility mode set at the registry level.

But another issue with this is that folks's code could manually be setting the compatibility mode differently for different subjects (the EventManager will do this), so if folks use this `PydanticSchemaManager directly, they'll have to explicitly write the code that says "if config.profile is development, set the subject compatibility to NONE".

It would be nice if there was something we can do in a CI run to check compatibility against prod or int, what do you think of doing something like that?

File: docs/user-guide/pydantic-schema-manager.rst
Comment (fajpunk) [reply docs/user-guide/pydantic-schema-manager.rst]: Ooh, `:caption:` looks nice, I didn't know about it!
image

Comment (fajpunk) [reply docs/user-guide/pydantic-schema-manager.rst]: ...but to you point about being prescriptive about specifying those attributes, I think that's a good idea, and I took out the other examples :smile: 
https://github.com/lsst-sqre/safir/pull/303/commits/695405aff7c64cc9d78bbadf792fafff012683b2

File: docs/user-guide/pydantic-schema-manager.rst
Comment (fajpunk) [reply docs/user-guide/pydantic-schema-manager.rst]: Ok, like we talked about in co-work, we're going to [kick a more robust solution to this down the road](https://rubinobs.atlassian.net/browse/DM-46529).  For now, I put a "don't screw this up" warning in the docs.

The metrics EventManager will set all subjects to be `FORWARD` compatible, which has been stable with EFD schema management usage. Any screw ups will be fixed manually for now :crossed_fingers:

**Review by rra — APPROVED** [review]
Looks good, thank you!

**Pull request merged**
